harness/determined/_generic.py

```diff
@@ -0,0 +1,259 @@
+                "workload": wkld,
+                "start_time": start_time,
+                "end_time": self._current_timestamp(),
+                "metrics": metrics,
+                "exited_reason": "INVALID_HP",
+            }
+            return out_response
+
+        if response.get("init_invalid_hp", False):
+            out_response = {
+                "type": "WORKLOAD_COMPLETED",
+                "workload": wkld,
+                "start_time": start_time,
+                "end_time": self._current_timestamp(),
+                "metrics": metrics,
+                "exited_reason": "INIT_INVALID_HP",
+            }
+            return out_response
+
+        v_metrics = metrics["validation_metrics"]
+        self._tensorboard_writer.on_validation_step_end(wkld.total_batches_processed, v_metrics)
+
+        self._tensorboard_mgr.sync()
+
+        # Check that the validation metrics computed by the model code
+        # includes the metric used by the search method.
+        if searcher_metric not in v_metrics:
+            raise AssertionError(
+                "Search method is configured to use metric '{}' but model "
+                "definition returned validation metrics {}. The metric "
+                "used by the search method must be one of the validation "
+                "metrics returned by the model definition.".format(
+                    searcher_metric, list(v_metrics.keys())
+                )
+            )
+
+        # Check that the searcher metric has a scalar value so that it can be compared for
+        # search purposes. Other metrics don't have to be scalars.
+        metric_value = v_metrics[searcher_metric]
+        if not tensorboard.metric_writers.util.is_numerical_scalar(metric_value):
+            raise AssertionError(
+                "Searcher validation metric '{}' returned "
+                "a non-scalar value: {}".format(searcher_metric, metric_value)
+            )
+
+        non_serializable_metrics = set()
+        # NaN and bytes are not JSON serializable. None does not have a
+        # canonical JSON representation. In the case of trial implementation bugs
+        # or numerical instability issues, validation metric functions may
+        # return None or NaN values. For now, immediately fail any trial that
+        # encounters such a None metric. For NaN metrics, if it's the target of
+        # the searcher, we set it to +/- max_float depending on if the searcher
+        # is optimizing for the max or min. NaN metrics which are not the
+        # target of the searcher are dropped.
+        # TODO (DET-2495): Do not replace NaN metric values.
+        for metric_name, metric_value in v_metrics.items():
+            metric_is_none = metric_value is None
+            metric_is_nan = tensorboard.metric_writers.util.is_numerical_scalar(
+                metric_value
+            ) and math.isnan(metric_value)
+
+            if metric_is_none or metric_is_nan:
```

Comment:
I think this behavior is "intentional", even though it seems poorly designed.  I did some forensics and this behavior has changed several times in opposing ways (`629f1068f` in `determined` and `ef53d3193` in `pedl`).  So this might not be the right behavior but it's been a problem frequently enough that I don't plan on changing the behavior in this PR.